### PR TITLE
[JENKINS-68933] Better WebSocket testing, removal of reflection

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-68933](https://issues.jenkins.io/browse/JENKINS-68933) and https://github.com/jenkinsci/jenkins/pull/6694#issuecomment-1174792020. Amending #4369.

### Proposed changelog entries

* N/A (internal changes only)

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6780"></a>

### CODE REVIEW

**Comment by jglick** [comment]
Looking over https://github.com/eclipse/jetty.project/pull/6623 & https://github.com/eclipse/jetty.project/pull/3544 & https://github.com/eclipse/jetty.project/pull/5874.

**Comment by jglick** [comment]
Rebased the test on top of the parent of #6694, where it passed.

**Comment by jglick** [comment]
Tests appear to fail here because the build is running PR-merge strategy. @timja @dduportal I thought we had decided to switch builds of this repo to use PR-head strategy? Maybe it was only `bom`?

**Comment by timja** [comment]
> Tests appear to fail here because the build is running PR-merge strategy. @timja @dduportal I thought we had decided to switch builds of this repo to use PR-head strategy? Maybe it was only `bom`?

only bom was switched

**Comment by jglick** [comment]
Split off #6785 to attempt (so far without success) to use this framework to support Jetty 10.

**Review by basil — APPROVED** [review]
Thanks Jesse, I had started working on a similar PR earlier today but you obviously got it done a lot faster. In the interest of saving time I took the liberty of pushing a few commits to your branch to fix static analysis issues. I also left a comment in https://github.com/jenkinsci/jenkins/pull/6785#discussion_r914305175 explaining how to get this to work with Jetty 10.

**Comment by basil** [comment]
Well now there's a a `javadoc: error - No public or protected classes found to document` but once that is resolved I think this is good to go.

**Review by timja — APPROVED** [review]

File: pom.xml
Comment (jglick) [inline pom.xml:99]: TIL

**Comment by jglick** [comment]
> In the interest of saving time

Well, there is no rush on this one since it does not fix anything for users, it is just making it easier to work on this code.

> I took the liberty of pushing a few commits to your branch to fix static analysis issues.

Thanks! I will check the Javadoc issue.

File: websocket/jetty9/src/main/java/jenkins/websocket/Jetty9Provider.java
Comment (jglick) [inline websocket/jetty9/src/main/java/jenkins/websocket/Jetty9Provider.java:115]: Note the `Object` in the signature. Javadoc is super helpful

> a websocket object to use

It is more or less through trial and error that I concluded (in #4369) that a `WebSocketListener` is expected; “normal usage” apparently means adding annotations to POJOs and just remembering what method signatures to use. I think this misconceived style is inherited from the `javax.websocket` API. (Which tries to make you declare WS-aware endpoints with more annotations, and fails to provide a simple Java API call to upgrade a connection, forcing us to rely on Jetty-specific APIs.)

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

File: bom/pom.xml
Comment (jglick) [inline bom/pom.xml:301]: Note that this will cause that library to be managed for plugin POMs (like `credentials` or `artifact-manager-s3`) for the first time. Seems like a good thing; I am not sure why it was not already in the core BOM. But if this is a problem for any reason, it could trivially be reverted, and the version fixed in the new module.

File: websocket/spi/pom.xml
Comment (jglick) [inline websocket/spi/pom.xml:55]: Deliberately not using BOM version here.

Comment (basil) [reply websocket/spi/pom.xml:55]: Yeah, I saw that and I agree with using 3.1.0 because that is the lowest common denominator for Jetty 9 and Jetty 10. But the downside is this is adding the JAR to `WEB-INF/lib/javax.servlet-api-3.1.0.jar` in the WAR, which I think is undesirable because it is not actually used at runtime; rather, Winstone's copy is used. So to fix this I think we should exclude it when we pull in `websocket-spi` in `core/pom.xml`, which already has a `provided` dependency on `servlet-api` that is aligned with Jetty's.

File: bom/pom.xml
Comment (basil) [reply bom/pom.xml:301]: FWIW there is an exclusion for this library elsewhere in the build: https://github.com/jenkinsci/jenkins/blob/644a261e78e9d80e2442787c8418cbdf283ac7e1/war/pom.xml#L92-L96=

File: websocket/jetty9/src/main/java/jenkins/websocket/Jetty9Provider.java
Comment (basil) [inline websocket/jetty9/src/main/java/jenkins/websocket/Jetty9Provider.java:51]: Is there any concern that a session might get dropped from the map while it is still active, resulting in `IllegalStateException("missing session")` later on? From a correctness perspective it seems better to explicitly remove the session from the map in both the successful termination and failed termination cases.

Comment (basil) [reply websocket/jetty9/src/main/java/jenkins/websocket/Jetty9Provider.java:51]: Never mind, as soon as I posted this I remembered that it's the keys of a `WeakHashMap` that are weak references, and the lifetime of the `Listener` key is tied to the `HttpServletRequest`. :thumbsup:

**Review by olamy — APPROVED** [review]
oh yes Jetty 10 API for websocket has changed a lot (it's supposed to be more easy to use but still a big change...)
good work here! (Hopefully we will not have to keep this jetty9 too long?)

File: websocket/jetty9/src/main/java/jenkins/websocket/Jetty9Provider.java
Comment (jglick) [inline websocket/jetty9/src/main/java/jenkins/websocket/Jetty9Provider.java:51]: regression: #7076

**Pull request merged**
